File: common-lib/src/main/java/club/hosppy/common/crypto/Sign.java
package club.hosppy.common.crypto;

import club.hosppy.common.api.ResultCode;
import club.hosppy.common.error.ServiceException;
import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.interfaces.DecodedJWT;
import org.apache.commons.lang3.StringUtils;

import java.util.Date;
import java.util.concurrent.TimeUnit;

public class Sign {

    public static final String CLAIM_EMAIL = "email";
    public static final String CLAIM_USER_ID = "userId";

    /**
     * a day
     */
    public static final long DURATION_MILLIS = 1000 * 60 * 60 * 24;

    private static Algorithm algorithm;

    public static String generateEmailConfirmationToken(String userId, String email, String secret) {
        instanceAlgorithm(secret);
        return JWT.create()
                .withClaim(CLAIM_EMAIL, email)
                .withClaim(CLAIM_USER_ID, userId)
                .withExpiresAt(new Date(System.currentTimeMillis() + TimeUnit.HOURS.toMillis(2)))
                .sign(algorithm);
    }

    public static String generateSessionToken(String userId, String secret) {
        if (StringUtils.isEmpty(secret)) {
            throw new ServiceException(ResultCode.NO_SIGNING_TOKEN);
        }
        instanceAlgorithm(secret);
        return JWT.create()
                .withClaim(CLAIM_USER_ID, userId)
                .withExpiresAt(new Date(System.currentTimeMillis() + DURATION_MILLIS))
                .sign(algorithm);
    }

    public static DecodedJWT verify(String token, String secret) {
        instanceAlgorithm(secret);
        JWTVerifier verifier = JWT.require(algorithm).build();
        return verifier.verify(token);
    }

    private static void instanceAlgorithm(String secret) {
        if (algorithm == null) {
            algorithm = Algorithm.HMAC512(secret);
        }
    }
}

File: src/main/kotlin/club/hosppy/common/crypto/Sign.kt
package club.hosppy.common.crypto

import club.hosppy.common.api.ResultCode
import club.hosppy.common.error.ServiceException
import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.interfaces.DecodedJWT
import org.apache.commons.lang3.StringUtils
import java.util.*
import java.util.concurrent.TimeUnit

object Sign {
    const val CLAIM_EMAIL = "email"
    const val CLAIM_USER_ID = "userId"

    /**
     * a day
     */
    private const val DURATION_MILLIS = (1000 * 60 * 60 * 24).toLong()
    private var algorithm: Algorithm? = null

    @JvmStatic
    fun generateEmailConfirmationToken(userId: String?, email: String?, secret: String): String {
        instanceAlgorithm(secret)
        return JWT.create()
            .withClaim(CLAIM_EMAIL, email)
            .withClaim(CLAIM_USER_ID, userId)
            .withExpiresAt(Date(System.currentTimeMillis() + TimeUnit.HOURS.toMillis(2)))
            .sign(algorithm)
    }

    fun generateSessionToken(userId: String?, secret: String): String {
        if (StringUtils.isEmpty(secret)) {
            throw ServiceException(ResultCode.NO_SIGNING_TOKEN)
        }
        instanceAlgorithm(secret)
        return JWT.create()
            .withClaim(CLAIM_USER_ID, userId)
            .withExpiresAt(Date(System.currentTimeMillis() + DURATION_MILLIS))
            .sign(algorithm)
    }

    @JvmStatic
    fun verify(token: String?, secret: String): DecodedJWT {
        instanceAlgorithm(secret)
        val verifier = JWT.require(algorithm).build()
        return verifier.verify(token)
    }

    private fun instanceAlgorithm(secret: String) {
        if (algorithm == null) {
            algorithm = Algorithm.HMAC512(secret)
        }
    }
}